src/classes/emitter.ts

```diff
@@ -1,49 +1,110 @@
+	addListener(eventName: EventName, listener: Listener): this;
+	off<T extends keyof Events>(
+		eventName: T,
+		listener: (this: this, ...args: Events[T]) => void,
+	): this;
+	off(eventName: EventName, listener: Listener): this;
+}
+
+type EventsOf<T> = T extends { _emitter: Emitter<infer U> } ? U
+	: T extends Emitter<infer U> ? U
+	: never;
+
+export function once<T extends Emitter, K extends keyof EventsOf<T>>(
+	emitter: T,
+	eventName: K,
+): Promise<EventsOf<T>[K]>;
+export function once(emitter: Emitter, eventName: EventName): Promise<any[]> {
+	return new Promise((res) => {
+		emitter.once(eventName, (...args) => res(args));
+	});
+}
+
+export default class Emitter<Events extends EventMap = EventMap> {
+	#events = new Map<EventName, Listener[]>();
+
+	static once = once;
+
+	static {
+		this.prototype.addListener = this.prototype.on;
+		this.prototype.off = this.prototype.removeListener;
+	}
+
+	// @ts-expect-error this is correct since we're overwriting with specific args.
+	on<T extends keyof Events>(
+		eventName: T,
+		listener: (this: this, ...args: Events[T]) => void,
+	): this;
+	on(eventName: EventName, listener: Listener): this {
+		const listeners = this.#events.get(eventName);
+		if (listeners !== undefined) {
+			listeners.push(listener);
+		} else {
+			this.#events.set(eventName, [listener]);
 		}
-		this.#events[e].push(func);
+		return this;
 	}
 
-	off(e: string, func: Listener) {
-		if (typeof this.#events[e] === "object") {
-			const idx = this.#events[e].indexOf(func);
+	// @ts-expect-error see above.
+	removeListener<T extends keyof Events>(
+		eventName: T,
+		listener: (this: this, ...args: Events[T]) => void,
+	): this;
+	removeListener(eventName: EventName, listener: Listener): this {
+		const events = this.#events.get(eventName);
+		if (events !== undefined) {
+			const idx = events.indexOf(listener);
 			if (idx > -1) {
-				this.#events[e].splice(idx, 1);
+				if (events.length === 1) {
+					this.#events.delete(eventName);
+				} else {
+					events.splice(idx, 1);
+				}
 			}
 		}
+		return this;
 	}
 
-	once(e: string, func: Listener) {
-		this.on(e, function f(...args) {
-			this.off(e, f);
-			func.apply(this, args);
+	// @ts-expect-error see above.
+	once<T extends keyof Events>(
+		eventName: T,
+		listener: (this: this, ...args: Events[T]) => void,
+	): this;
+	once(eventName: EventName, listener: Listener): this {
+		this.on(eventName, function once(...args: any[]) {
+			this.removeListener(eventName, listener);
+			listener.apply(this, args);
 		});
+		return this;
 	}
 
-	emit(e: string, ...args: any[]) {
-		if (typeof this.#events[e] === "object") {
-			this.#events[e].forEach((func) => {
-				func.apply(this, args);
-			});
+	emit<T extends keyof Events>(eventName: T, ...args: Events[T]): this;
+	emit(eventName: EventName, ...args: any[]): this {
+		const listeners = this.#events.get(eventName);
+		if (listeners !== undefined) {
+			listeners.forEach((listener) => listener.apply(this, args));
```

Comment:
why use apply. This will break a lot of code I think `listener(...args)` would be better.
But if we leave it yes we would need the this evrywhere....